src/modules/replay/replay_main.cpp

```diff
@@ -752,11 +731,273 @@ void Replay::task_main()
+						advertised = true;
+					}
+				}
+			}
+
+			if (advertised) {
+				int instance;
+				sub.orb_advert = orb_advertise_multi(sub.orb_meta, data, &instance, ORB_PRIO_DEFAULT);
+				published = true;
+			}
+		}
+	}
+
+	if (published) {
+		++sub.publication_counter;
+	}
+
+	return published;
+}
+
+bool ReplayEkf2::handleTopicUpdate(Subscription &sub, void *data, std::ifstream &replay_file)
+{
+	if (sub.orb_meta == ORB_ID(ekf2_timestamps)) {
+		ekf2_timestamps_s ekf2_timestamps;
+		memcpy(&ekf2_timestamps, data, sub.orb_meta->o_size);
+
+		if (!publishEkf2Topics(ekf2_timestamps, replay_file)) {
+			return false;
+		}
+
+		px4_pollfd_struct_t fds[1];
+		fds[0].fd = _vehicle_attitude_sub;
+		fds[0].events = POLLIN;
+		// wait for a response from the estimator
+		int pret = px4_poll(fds, 1, 1000);
+
+		// introduce some breaks to make sure the logger can keep up
+		if (++_topic_counter == 50) {
+			usleep(1000);
+			_topic_counter = 0;
+		}
+
+		if (pret == 0) {
+			PX4_WARN("poll timeout");
+
+		} else if (pret < 0) {
+			PX4_ERR("poll failed (%i)", pret);
+
+		} else {
+			if (fds[0].revents & POLLIN) {
+				vehicle_attitude_s att;
+				// need to to an orb_copy so that poll will not return immediately
+				orb_copy(ORB_ID(vehicle_attitude), _vehicle_attitude_sub, &att);
+			}
+		}
+
+		return true;
+
+	} else if (sub.orb_meta == ORB_ID(vehicle_status) || sub.orb_meta == ORB_ID(vehicle_land_detected)) {
+		return publishTopic(sub, data);
+	} // else: do not publish
+
+	return false;
+}
+
+void ReplayEkf2::onSubscriptionAdded(Subscription &sub, uint16_t msg_id)
+{
+	if (sub.orb_meta == ORB_ID(sensor_combined)) {
+		_sensors_combined_msg_id = msg_id;
+
+	} else if (sub.orb_meta == ORB_ID(vehicle_gps_position)) {
+		_gps_msg_id = msg_id;
+
+	} else if (sub.orb_meta == ORB_ID(optical_flow)) {
+		_optical_flow_msg_id = msg_id;
+
+	} else if (sub.orb_meta == ORB_ID(distance_sensor)) {
+		_distance_sensor_msg_id = msg_id;
+
+	} else if (sub.orb_meta == ORB_ID(airspeed)) {
+		_airspeed_msg_id = msg_id;
+
+	} else if (sub.orb_meta == ORB_ID(vehicle_vision_position)) {
+		_vehicle_vision_position_msg_id = msg_id;
+
+	} else if (sub.orb_meta == ORB_ID(vehicle_vision_attitude)) {
+		_vehicle_vision_attitude_msg_id = msg_id;
+	}
+
+	// the main loop should only handle publication of the following topics, the sensor topics are
+	// handled separately in publishEkf2Topics()
+	sub.ignored = sub.orb_meta != ORB_ID(ekf2_timestamps) && sub.orb_meta != ORB_ID(vehicle_status)
+		      && sub.orb_meta != ORB_ID(vehicle_land_detected);
+}
+
+bool ReplayEkf2::publishEkf2Topics(const ekf2_timestamps_s &ekf2_timestamps, std::ifstream &replay_file)
+{
+	auto handle_sensor_publication = [&](int16_t timestamp_relative, uint16_t msg_id) {
+		if (timestamp_relative != ekf2_timestamps_s::RELATIVE_TIMESTAMP_INVALID) {
+			uint64_t t = timestamp_relative + ekf2_timestamps.timestamp / 100; // in 0.1 ms
```

Comment:
`timestamp_relative` is in 0.1 ms as well and to compare it we need to use the same unit. I chose 0.1 ms because it allows to use an int16 (keep logging rate low), while at the same time allowing sensor publications up to 10kHz.